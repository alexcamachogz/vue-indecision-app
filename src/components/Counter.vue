<template>
  <h2>{{ customTitle }}</h2>
  <p>{{ counter }} <sup>2</sup> = {{ squareCounter }}</p>

  <div>
    <button @click="increase()"> + </button>
    <button @click="decrease()"> - </button>
  </div>
</template>

<script>
export default {
  props: {
    title: String,
    start: {
      type: Number,
      default: 100,
      validator (value) {
        return value >= 100
      }
    }
  },
  data () {
    return {
      counter: this.start
    }
  },
  methods: {
    // increase counter
    increase () {
      this.counter++
    },
    // decrease counter
    decrease () {
      this.counter--
    }
  },
  computed: {
    squareCounter () {
      return this.counter * this.counter
    },
    customTitle () {
      return this.title || 'Counter'
    }
  }
}
</script>

<style>
  button {
    background-color: #64B687;
    border-radius: 5px;
    border: 1px solid white;
    cursor: pointer;
    color: white;
    margin: 0 5px;
    padding: 5px 15px;
    transition: 0.3s ease-in-out;
    width: 60px;
    height: 35px;
  }

  button:hover {
    background-color: #5AA67B;
    transition: 0.3s ease-in-out;
  }
</style>
