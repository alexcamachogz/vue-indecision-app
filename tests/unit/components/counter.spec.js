import { shallowMount } from '@vue/test-utils'
import Counter from '@/components/Counter'

describe('Counter component', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(Counter)
  })

  test('Should be match with the snapshot', () => {
    const wrapper = shallowMount(Counter)
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render value by default', () => {
    expect(wrapper.find('h2').exists()).toBeTruthy()

    const h2Value = wrapper.find('h2').text()
    expect(h2Value).toEqual('Counter')
  })

  test('should be 100 the defaul value on p', () => {
    const pValue = wrapper.find('[data-testId="counter"]').text()
    expect(pValue).toBe('100')
  })

  test('should be increase and decrease in 1 the counter value', async () => {
    const [incraseBtn, decreaseBtn] = wrapper.findAll('button')

    await incraseBtn.trigger('click')
    await incraseBtn.trigger('click')
    await incraseBtn.trigger('click')
    await decreaseBtn.trigger('click')
    await decreaseBtn.trigger('click')

    const value = wrapper.find('[data-testId="counter"]').text()
    expect(value).toBe('101')
  })

  test('should be set the default value', () => {
    const start = wrapper.props('start')
    const value = wrapper.find('[data-testId="counter"]').text()
    expect(Number(value)).toBe(start)
  })

  test('should be show the prop title', () => {
    const title = 'Hola Alex'
    const wrapper = shallowMount(Counter, {
      props: {
        title
      }
    })
    const value = wrapper.find('h2').text()
    expect(value).toBe(title)
  })
})
