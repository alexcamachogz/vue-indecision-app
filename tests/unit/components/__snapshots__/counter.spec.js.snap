// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component Should be match with the snapshot 1`] = `
<h2>Counter</h2>
<p>100 <sup>2</sup> = 10000</p>
<p data-testid="counter">100</p>
<div class="btn-container"><button> + </button><button> - </button></div>
`;
